### PR TITLE
update Unet3d

### PR DESCRIPTION
**Pull request by suexu1025**
- add unet3d support for brats data
- fix data loading slow on gsc using long file path
- add run_brats/run_kitts scripts

UNET3D @ BRATS with BS 64 convergent at mean epoch of 63.7 with std 9.2 @0.845 DICE
UNET3D @ KITS with BS 56 convergent at mean epoch of 2213.4 with std 499.9. @0.908 DICE
e2e training time of BRATS vs KITS ~ 1:4

general performance from this PR:
on v4-8 expected throughputs of 17.8 samples/s on GCS.
Improvement over original impl: 28% using disk, 48% using GCS
on v4-128 expected throughputs of 263.2 samples/s on GCS
Improvement over original impl: 2.7x using disk 6.4x using GCS.

### CODE REVIEW

**Comment by JackCaoG** [comment]
I don't have enough context to review the full pr, @miladm Do you want to give it a try?

File: image_segmentation/pytorch/main.py [hunk 1]
```diff
@@ -68,8 +69,12 @@ def main(local_rank, flags):
         mlperf_run_param_log(flags)
 
     callbacks = get_callbacks(flags, dllogger, local_rank, world_size)
+
+    if flags.use_brats:
```
Comment (miladm) [inline image_segmentation/pytorch/main.py:73]: I'd say we can have Brats as a use-case example. To given the state of Unet3D in MLPerf, it makes sense to keep Brats as an optional choice.

This means your PR focuses on Unet3D general optimizations, and enables those interested in Brats to give it a try.

File: image_segmentation/pytorch/main.py [hunk 2]
```diff
@@ -152,13 +153,19 @@ def patched_optimizer_step(optimizer, barrier=False, optimizer_args={}):
...
     )
     score_fn = DiceScore(
         to_onehot_y=True,
         use_argmax=True,
         layout=flags.layout,
         include_background=flags.include_background,
+        num_classes=4 if flags.use_brats else 3
     )
+    
+    # to match original paper, but still high variance
```
Comment (miladm) [inline image_segmentation/pytorch/main.py]: do you need these comments to stay?

File: image_segmentation/pytorch/model/losses.py [hunk 3]
```diff
@@ -55,25 +57,25 @@ def __call__(self, prediction, target):
 
         intersection = torch.sum(target * prediction, dim=reduce_axis)
         target_sum = torch.sum(target, dim=reduce_axis)
-        prediction_sum = torch.sum(prediction, dim=reduce_axis)
+        prediction_sum = torch.sum(prediction, dim=reduce_axis).float()
```
Comment (miladm) [inline image_segmentation/pytorch/model/losses.py]: why do you need this `float()` conversion?

File: image_segmentation/pytorch/runtime/inference.py [hunk 4]
```diff
@@ -36,43 +39,60 @@ def evaluate(flags, model, loader, loss_fn, score_fn, device, epoch=0, is_distri
...
-                    labels=label,
-                    roi_shape=flags.val_input_shape,
-                    model=model,
-                    overlap=flags.overlap,
-                    mode="gaussian",
-                    padding_val=-2.2
-                )
-                eval_loss_value = loss_fn(output, label)
-                scores.append(score_fn(output, label))
+            #with autocast(enabled=flags.amp):
```
Comment (miladm) [inline image_segmentation/pytorch/runtime/inference.py]: do we not need amp `autocast`? if not, do you plan to drop this line?

File: image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py [hunk 5]
```diff
@@ -103,8 +105,15 @@ def train(self):
 
                     loss_value = self.train_step(iteration=iteration, images=images, labels=labels)
 
-                    #loss_value = reduce_tensor(loss_value).detach().cpu().numpy()
+                    loss_value = reduce_tensor(loss_value)
                     cumulative_loss.append(loss_value)
+                    if self.summary_writer:
+                        test_utils.write_to_summary(
+                            self.summary_writer,
+                            global_step=epoch * 10000 // 64 + iteration,
```
Comment (miladm) [inline image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py]: do you need to keep these numbers hard-coded?

File: image_segmentation/pytorch/runtime/trainer/xla_trainer.py [hunk 6]
```diff
@@ -49,6 +50,26 @@ def __init__(
...
+        
+        if flags.tb_dir !="":
+            dataset_name = "kitts/" if not flags.use_brats else ""
+            self.summary_dir = flags.tb_dir + dataset_name + datetime.now().strftime("%Y%m%d-%H%M%S")
+            self.summary_interval = 100
+            if xm.is_master_ordinal():
+                self.summary_writer = test_utils.get_summary_writer(
+                    self.summary_dir) if self.summary_interval else None
+
+            # test_utils.write_to_summary(
```
Comment (miladm) [inline image_segmentation/pytorch/runtime/trainer/xla_trainer.py]: do you need these commented lines?

**Review by miladm — CHANGES_REQUESTED** [review]
Thanks for the PR @suexu1025 

Left a few questions and comments here.

We need to outline the perf gains w/ and w/o Brats in this PR. Feel free to make a separate PR for Brats if more convenient - your call.

File: image_segmentation/pytorch/main.py (hunk 1 above)
Comment (suexu1025) [reply image_segmentation/pytorch/main.py:73]: sg, will keep brats as optional.

File: image_segmentation/pytorch/runtime/inference.py (hunk 4 above)
Comment (suexu1025) [reply image_segmentation/pytorch/runtime/inference.py]: restored

**Comment by suexu1025** [comment]
# address PR comment and summary of the PR benchmark result.

UNET3D @ BRATS with BS 64 convergent at mean epoch of  63.7 with std 9.2 @0.845 DICE
UNET3D @ KITS with BS 56 convergent at mean epoch of  2213.4 with std 499.9. @0.908 DICE
e2e training time of BRATS vs KITS ~ 1:4

general performance from this PR:
on v4-8 expected throughputs of 17.8 samples/s on GCS.
Improvement over original impl: 28% using disk, 48% using GCS
on v4-128 expected throughputs of 263.2 samples/s on GCS 
Improvement over original impl: 2.7x using disk 6.4x using GCS.

**Pull request merged**